### PR TITLE
Add restoration from backup file to a zxid restoration point

### PR DESCRIPTION
**Pull request by mgao0**
This commit adds a tool called RestoreFromBackupTool to allow restoration of snapshots and transaction logs to a certain zxid restoration point. This tool is going to be used together with a ZK CLI where user can input commands to restore.

New classes:
RestoreFromBackupTool, RestoreException, RestorationToolTest
Tests:
Added tests RestorationToolTest. More local testing to be done once ZK CLI is completed.

### CODE REVIEW

File: zookeeper-server/pom.xml [hunk 1]
```diff
@@ -167,6 +167,12 @@
       <artifactId>guava</artifactId>
       <version>27.0.1-jre</version>
     </dependency>
+    <dependency>
+      <groupId>commons-io</groupId>
+      <artifactId>commons-io</artifactId>
+      <version>2.6</version>
+      <scope>compile</scope>
+    </dependency>
```
Comment (abhilash1in) [inline zookeeper-server/pom.xml]: Do we need a new dependency just for using `FileUtils.listFiles()` and `FileUtils.deleteDirectory()`? Can we implement this ourselves?

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/storage/impl/FileSystemBackupStorage.java [hunk 2]
```diff
@@ -184,8 +186,15 @@ public void copyToLocalStorage(File srcName, File destFile) throws IOException {
...
-    String backupFilePath =
-        BackupStorageUtil.constructBackupFilePath(srcName.getName(), fileRootPath);
+    String parentDirPath;
+    String srcFileName = srcName.getName();
+    if (srcFileName.startsWith(BackupStorageUtil.RESTORE_FILE_PREFIX)) {
+      parentDirPath = srcName.getParent();
+      srcFileName = srcFileName.substring(BackupStorageUtil.RESTORE_FILE_PREFIX.length());
+    } else {
+      parentDirPath = fileRootPath;
+    }
```
Comment (junkaixue) [inline zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/storage/impl/FileSystemBackupStorage.java]: NIT: else clause is not necessary.

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/RestoreFromBackupTool.java [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+ *  that combine to represent a transactionally consistent image of a ZooKeeper ensemble at
+ *  some zxid.
+ */
+public class RestoreFromBackupTool {
+  private static final Logger LOG = LoggerFactory.getLogger(RestoreFromBackupTool.class);
+
+  BackupStorageProvider storage;
+  FileTxnSnapLog snapLog;
+  long zxidToRestore;
+  boolean dryRun;
```
Comment (junkaixue) [inline zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/RestoreFromBackupTool.java:51]: Question: Is this for stateful tool? My understanding this is a class will be hooked by shell script and call every time. Do we need a class level variable for that? Shall we make all the method as stateless or static method?

File: zookeeper-server/pom.xml (hunk 1 above)
Comment (mgao0) [reply zookeeper-server/pom.xml]: I think your concern is valid. I updated the code to remove usage of FileUtils.

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/RestoreFromBackupTool.java [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+package org.apache.zookeeper.server.backup;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Range;
+import jline.internal.Log;
```
Comment (narendly) [inline zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/RestoreFromBackupTool.java]: Do we need this import?

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/RestoreFromBackupTool.java (hunk 3 above)
Comment (mgao0) [reply zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/RestoreFromBackupTool.java:51]: This class is stateful. Every time the command is triggered, we instantiate an instance of this class, and parse the arguments and store them as variables. In my next PR, the CLI command will be implemented in the same class, making the class stateful enable us to clearly separate the CLI command implementation and the restoration logic.

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/RestoreFromBackupTool.java [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+
+          if (!dryRun) {
+            // Adding "RESTORE_" prefix to file name to differentiate the restore files from original backup files,
+            // so in storage side it will use a different way to construct the file path from original backup files
+            storage.copyToLocalStorage(
+                new File(parentFile, BackupStorageUtil.RESTORE_FILE_PREFIX + fileName),
+                destination);
+          }
+        } else {
+          LOG.info("Skipping " + fileName + " because it already exists as " + destination.getPath()
```
Comment (narendly) [inline zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/RestoreFromBackupTool.java]: Nit: Skipping "copying" fileName because it...?

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/RestoreFromBackupTool.java [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+              + ".");
+        }
+      }
+    } catch (IOException ioe) {
+      throw new IOException("Hit exception when attempting to restore." + ioe.getMessage());
```
Comment (narendly) [inline zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/RestoreFromBackupTool.java]: Do we really need to catch and rethrow here since it's the same IOException? Should we just let the original exception be thrown?

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/RestoreFromBackupTool.java [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+        File parentFile = processedFile.getParentFile();
+        File base = Util.isSnapshotFileName(fileName) ? snapLog.getSnapDir() : snapLog.getDataDir();
+        File destination = new File(base, fileName);
+
+        if (!destination.exists()) {
+          LOG.info("Copying " + fileName + " to " + destination.getPath() + ".");
+
+          if (!dryRun) {
+            // Adding "RESTORE_" prefix to file name to differentiate the restore files from original backup files,
+            // so in storage side it will use a different way to construct the file path from original backup files
```
Comment (narendly) [inline zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/RestoreFromBackupTool.java]: This sounds like as if you are adding the prefix to the files you're copying to (files created in local storage as a result of copying). I don't believe this is what you really meant - did you mean that the source files (already) _have_ the RESTORE prefix prepended to differentiate in the _backup (not local) storage_?

Is this differentiation necessary though? We are already differentiating by putting them in a temp directory so we don't mix them up with the original backup files, right?

Perhaps another thing I might want to do here is to add/modify prefixes to differentiate files before/after processing?

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/storage/BackupStorageProvider.java [hunk 8]
```diff
@@ -95,4 +95,15 @@
...
+
+  /**
+   * Copy the selected snapshot and transaction log backup files needed for restoration to a temp folder;
+   * Truncate the copied transaction log to zxid restoration point.
+   * @param tempDirDestination The path to create a temporary directory and store the copied backup files
+   * @param filesToCopy A list of selected backup files that can be used for restoration
+   * @param zxidToRestore The zxid to restore to
+   * @throws IOException
+   */
+  void processBackupFilesForRestoration(File tempDirDestination, List<BackupFileInfo> filesToCopy,
```
Comment (narendly) [inline zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/storage/BackupStorageProvider.java]: I'm not sure if this is the right place for this processing logic (interface) to be in. In my opinion, Storage Provider interface should strictly provide I/O interface methods - restoration logic should go to a different place - either a separate interface, or part of the restoration tool? Do you think we should move this logic elsewhere?

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/exception/RestoreException.java [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.zookeeper.server.backup.exception;
+
+/**
+ * Signals an out-of-range error has occurred during the ZooKeeper restore process.
+ * For example: User requested to restore to a timestamp earlier than the earliest timestamp recorded in timetable,
```
Comment (narendly) [inline zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/exception/RestoreException.java]: This usage (timestamp being out of range) is currently represented by an IllegalArgumentException in TimetableUtil. I think either is fine, but if we want to introduce another exception, we should update `TimetableUtil` just to keep things consistent. There are other modes of failure in `TimetableUtil` that could potentially use `RestoreException` as well, so if we want to include those failure modes, we should update accordingly in this PR.

I am personally not entirely sure if this extra exception is needed. Perhaps it's a good idea to use this for restoration-related logic (other than IllegalArgument, IOException, etc.), but one thing I would encourage is to change the scope of the out-of-range error because for that particular error case, I think IllegalArgument might be more appropriate? What do you think?

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/RestoreFromBackupTool.java [hunk 10]
```diff
@@ -0,0 +1,295 @@
...
+   * Attempts to perform a restore.
+   */
+  public void run() throws IOException {
+    if (!findFilesToRestore()) {
+      throw new RestoreException("Failed to find a valid snapshot and logs to restore.");
+    }
+
+    // Create a temporary dir in backup storage for backup file processing
+    File tempDirForRestoration =
+        new File(filesToCopy.listIterator().next().getBackedUpFile().getParent(),
```
Comment (narendly) [inline zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/RestoreFromBackupTool.java]: Should we create a local variable called `restorationTmpDir`? and compute it during `findFilesToRestore()`? 
Alternatively, we could expose (add a getter) the storage path in BackupStorageProvider.

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/RestoreFromBackupTool.java [hunk 11]
```diff
@@ -71,11 +76,12 @@ public RestoreFromBackupTool() {
    * @param zxidToRestore the zxid upto which to restore, or Long.MAX to restore to the latest
    *                      available transactionally consistent zxid.
    * @param dryRun whether this is a dryrun in which case no files are actually copied
+   * @param restoreTempDir a local temporary directory to store the copied files from backup storage for restoration
```
Comment (narendly) [inline zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/RestoreFromBackupTool.java]: We can reword this to:

> a temporary, local (not remote) directory to stage the backup files from remote backup storage for the processing stage of restoration. Note that this directory and the files in it will be removed after restoration.

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/RestoreFromBackupTool.java [hunk 12]
```diff
@@ -84,52 +90,50 @@ public RestoreFromBackupTool() {
...
   /**
    * Attempts to perform a restore.
    */
   public void run() throws IOException {
-    if (!findFilesToRestore()) {
-      throw new RestoreException("Failed to find a valid snapshot and logs to restore.");
-    }
+    try {
+      if (!findFilesToRestore()) {
+        throw new IllegalArgumentException("Failed to find a valid snapshot and logs to restore.");
```
Comment (narendly) [inline zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/RestoreFromBackupTool.java]: Nit: remove "a"

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/RestoreFromBackupTool.java [hunk 13]
```diff
@@ -84,52 +90,50 @@ public RestoreFromBackupTool() {
...
+      }
 
-    // Create a temporary dir in backup storage for backup file processing
-    File tempDirForRestoration =
-        new File(filesToCopy.listIterator().next().getBackedUpFile().getParent(),
-            "Restore_" + Long.toHexString(zxidToRestore));
+      if (snapLog == null) {
+        throw new BackupException(
+            "The destination directory for restoration are not specified, please check the input.");
+      }
```
Comment (narendly) [inline zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/RestoreFromBackupTool.java]: `FileTxnSnapLog` isn't strictly a directory per se, so I'm not sure if this check is necessary. If you want to keep this check, something like "FileTxnSnapLog cannot be null!" would be more appropriate - what do you think?

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/RestoreFromBackupTool.java [hunk 14]
```diff
@@ -84,52 +90,50 @@ public RestoreFromBackupTool() {
...
-          }
-        } else {
-          LOG.info("Skipping " + fileName + " because it already exists as " + destination.getPath()
-              + ".");
+      if (!snapLog.getDataDir().exists() || !snapLog.getSnapDir().exists()) {
+        throw new BackupException(
+            "The destination directories for restoration do not exist, please check the input. Destination path for snapshots: "
+                + snapLog.getSnapDir().getPath() + ", destination path for txn logs: " + snapLog
+                .getDataDir().getPath());
+      }
```
Comment (narendly) [inline zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/RestoreFromBackupTool.java]: Do they need to exist first, or rather, why do we need to make this a requirement? 

Or can we just create if they do not exist already? I am thinking perhaps we could do the latter because we might want to support manual bootstrapping of a server before it joins the quorum node (this is to reduce the burden on the LEADER node), and if that's the case, the server may not have these directories already created.

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/RestoreFromBackupTool.java [hunk 15]
```diff
@@ -0,0 +1,373 @@
...
+
+      if (!snapLog.getDataDir().exists() || !snapLog.getSnapDir().exists()) {
+        throw new BackupException(
+            "The destination directories for restoration do not exist, please check the input. Destination path for snapshots: "
+                + snapLog.getSnapDir().getPath() + ", destination path for txn logs: " + snapLog
+                .getDataDir().getPath());
+      }
+
+      if (!dryRun) {
+        if (!restoreTempDir.exists() && !restoreTempDir.mkdirs()) {
```
Comment (narendly) [inline zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/RestoreFromBackupTool.java:121]: If you want to do validation on the arguments passed in, we should do a null check on `restoreTmpDir` as well.

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/RestoreFromBackupTool.java [hunk 16]
```diff
@@ -0,0 +1,373 @@
...
+
+    return true;
+  }
+
+  /**
+   * Copy selected backup files from backup storage to a local restore temporary directory for further processing later
+   * @param restoreTempSnapLog A FileTxnSnapLog instance created on the specified local temporary directory path
+   * @throws IOException
+   */
+  private void copyRestoredFilesToLocalTempDir(FileTxnSnapLog restoreTempSnapLog)
```
Comment (narendly) [inline zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/RestoreFromBackupTool.java]: Nit: `copyBackupFilesToLocalTempDir` sounds better?

I'm not sure if `restored` is the right way to describe these files because to me, `restored` means all processing/moving are complete (aka restoration is complete). In this case, we are referring to the backup files prior to any processing.

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/RestoreFromBackupTool.java [hunk 17]
```diff
@@ -0,0 +1,373 @@
...
+      throws IOException {
+    // Copy backup files to local temp directory
+    for (BackupFileInfo backedupFile : filesToCopy) {
+      String standardFilename = backedupFile.getStandardFile().getName();
+      // Does not matter if it's dataDir or logDir since we use same path for these two directories
```
Comment (narendly) [inline zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/RestoreFromBackupTool.java]: What did you mean by this? This comment is confusing because I think it's possible that dataDir and logDir are different for certain systems? I think what you meant is that we just make them the same for the local temp directory FileSnapLog object for restoration. Let's clarify this.

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/RestoreFromBackupTool.java [hunk 18]
```diff
@@ -0,0 +1,373 @@
...
+      File localTempDest = new File(restoreTempSnapLog.getDataDir(), standardFilename);
+
+      if (!localTempDest.exists()) {
+        LOG.info("Copying " + backedupFile.getBackedUpFile() + " from backup storage to temp dir "
+            + localTempDest.getPath() + ".");
+        storage.copyToLocalStorage(backedupFile.getBackedUpFile(), localTempDest);
+      } else {
+        LOG.info(
+            "Skipping copying " + backedupFile.getBackedUpFile() + " because it already exists as "
+                + localTempDest.getPath() + ".");
```
Comment (narendly) [inline zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/RestoreFromBackupTool.java]: This seems like an error case to me - is it not? Should we make it an error log? Under what circumstances would this happen and would this be acceptable? (if not, we should throw an exception or something here)

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/RestoreFromBackupTool.java [hunk 19]
```diff
@@ -0,0 +1,373 @@
...
+            .collect(Collectors.toList());
+    for (File processedFile : processedFiles) {
+      String fileName = processedFile.getName();
+      File finalDestinationBase = null;
+      if (Util.isSnapshotFileName(fileName)) {
+        finalDestinationBase = snapLog.getSnapDir();
+      } else if (Util.isLogFileName(fileName)) {
+        finalDestinationBase = snapLog.getDataDir();
+      }
+      if (finalDestinationBase != null) {
```
Comment (narendly) [inline zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/RestoreFromBackupTool.java]: If `finalDestinationBase` is `null`, do we just do nothing? Or can it ever be null? I might be misunderstanding it, but I feel like we could tighten the logic here a little more.

Comment (mgao0) [reply zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/RestoreFromBackupTool.java]: It does not mean something went wrong - the only case I can think of (but actually exists) is that "initialize" file being created by `ClientBase.createTmpDir();` that we used in the tests, and that's why I just skip this file and do nothing here. Let me sync up offline with you and let's brainstorm if there are other cases for this situation.

Comment (narendly) [reply zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/RestoreFromBackupTool.java]: In that case,  you could do the check for isSnapshot or isLog, if neither, you can just simply `continue` :)

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/RestoreFromBackupTool.java (hunk 7 above)
Comment (mgao0) [reply zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/RestoreFromBackupTool.java]: Discussed offline, this comment along with the next a few comments will be addressed by creating the temp dir in local instead of in backup storage. Creating temp dir in backup storage requires one implementation for each backup storage type (GPFS, HDFS, etc.) and will need special handlings on the files, which is a heavy-weighted operation. Therefore we decided to creating the temp dir in local instead which one implementation can be used for all the backup storages, and does not require special handling like this.

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/RestoreFromBackupTool.java (hunk 15 above)
Comment (mgao0) [reply zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/RestoreFromBackupTool.java:121]: I refactored the code a bit, moved all null/validation checks to the beginning, and cleaned up the existing checks. This should address this one and the above comments about validation checks.

File: zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/RestoreFromBackupTool.java [hunk 20]
```diff
@@ -0,0 +1,375 @@
...
+        FileTxnSnapLog restoreTempSnapLog =
+            new FileTxnSnapLog(this.restoreTempDir, this.restoreTempDir);
+
+        copyBackupFilesToLocalTempDir(restoreTempSnapLog);
+        processCopiedBackupFiles(restoreTempSnapLog, zxidToRestore);
+        copyProcessedRestoredFilesToDestination(restoreTempSnapLog);
+      }
+    } finally {
+      if (restoreTempDir != null && restoreTempDir.exists()) {
+        BackupStorageUtil.deleteDirectoryRecursively(restoreTempDir);
```
Comment (narendly) [inline zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/RestoreFromBackupTool.java:139]: Nit: let's add a quick comment here why we need to delete recursively - this is because "version-2" gets created with FileTxnSnapLog.

**Review by narendly — APPROVED** [review]
